File: iter1/322.cpp
//
// Created by chenyufan on 2020/5/12.
//

#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0) return 0;

        vector<unsigned int> f(amount+2, -1);
        f[0] = 0;

        for (int i = 1; i <= amount; ++i) {
            for (int coin : coins) {
                if (i - coin >= 0) {
                    f[i] = min(f[i-coin] + 1, f[i]);
                }
            }
        }
        return f[amount];
    }

};


int main() {
    Solution s;
    vector<int> c{1,2,5};
    cout << s.coinChange(c, 1) << endl;
}

File: iter2/322.py
from functools import lru_cache

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @lru_cache(None)
        def dfs(amount):
            if amount == 0: return 0
            ans = min([dfs(amount-i) for i in coins if i <= amount], default=float('inf'))
            return ans+1 if ans != float('inf') else ans
        return ans if ans != float('inf') else -1